test/Npgsql.Tests/Types/JsonDynamicTests.cs

```diff
@@ -215,87 +215,92 @@ await AssertType(
-            isDefault: false);
+        await AssertTypeWrite(dataSource, value, sql, PostgresType, NpgsqlDbType, isNpgsqlDbTypeInferredFromClrType: false);
+        await AssertTypeRead<WeatherForecast>(dataSource, sql, PostgresType, value, isDefault: false);
     }
 
     [Test]
     public async Task Poco_polymorphic_mapping_read_parents()
     {
-        var dataSourceBuilder = CreateDataSourceBuilder()
-            .EnableDynamicJson(jsonClrTypes: [typeof(WeatherForecast)])
-            .ConfigureJsonOptions(new () { AllowOutOfOrderMetadataProperties = true });
-        await using var dataSource = dataSourceBuilder.Build();
+        await using var dataSource = CreateWeatherForecastDataSource();
 
-        var value = new ExtendedDerivedWeatherForecast()
+        var value = new ExtendedDerivedWeatherForecast
         {
             Date = new DateTime(2019, 9, 1),
             Summary = "Partly cloudy",
             TemperatureC = 10
         };
 
+        // Note: we assert a specific string representation, though jsonb doesn't guarantee the property ordering; so the assert may break
+        // for jsonb if PostgreSQL changes its implementation.
         var sql =
             IsJsonb
                 ? """{"Date": "2019-09-01T00:00:00", "$type": "extended", "Summary": "Partly cloudy", "TemperatureC": 10, "TemperatureF": 49}"""
                 : """{"$type":"extended","TemperatureF":49,"Date":"2019-09-01T00:00:00","TemperatureC":10,"Summary":"Partly cloudy"}""";
 
-        await AssertTypeWrite<WeatherForecast>(
-            dataSource,
-            value,
-            sql,
-            PostgresType,
-            NpgsqlDbType,
-            isNpgsqlDbTypeInferredFromClrType: false,
-            isDefault: false);
+        await AssertTypeWrite<WeatherForecast>(dataSource, value, sql, PostgresType, NpgsqlDbType,
+            isNpgsqlDbTypeInferredFromClrType: false);
 
-        // GetFieldValue
-        await AssertTypeRead<WeatherForecast>(dataSource, sql, PostgresType, value,
-            comparer: (_, actual) => actual.GetType() == typeof(ExtendedDerivedWeatherForecast),
+        await AssertTypeRead<WeatherForecast>(dataSource, sql, PostgresType, value, isDefault: false);
+        await AssertTypeRead(dataSource, sql, PostgresType,
+            new DerivedWeatherForecast
+            {
+                Date = new DateTime(2019, 9, 1),
+                Summary = "Partly cloudy",
+                TemperatureC = 10
+            },
             isDefault: false);
+        await AssertTypeRead(dataSource, sql, PostgresType, value, isDefault: false);
+    }
 
-        await AssertTypeRead<DerivedWeatherForecast>(dataSource, sql, PostgresType, value,
-            comparer: (_, actual) => actual.GetType() == typeof(DerivedWeatherForecast), isDefault: false);
+    [Test]
+    public async Task Poco_exact_polymorphic_mapping()
+    {
+        await using var dataSource = CreateWeatherForecastDataSource();
```

Comment:
@roji factoring out the datasource creation introduced a bug, none of the outputs should contain $type. It's probably best to keep EnableDynamicJson in the tests as the clrtypes value are absolutely relevant to the test data.

Reply:
You could pass the type(s) in as an argument to that helper method